cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -115,6 +118,94 @@ def _decompose_(self):
             hamiltonian_polynomial_list, self._qubit_map, self._theta
         )
 
+    def _has_unitary_(self):
+        return True
+
+    def __repr__(self):
+        return (
+            f'cirq.BooleanHamiltonian('
+            f'qubit_map={self._qubit_map!r}, '
+            f'boolean_strs={self._boolean_strs!r}, '
+            f'theta={self._theta!r})'
+        )
+
+
+@value.value_equality
+class BooleanHamiltonianGate(raw_types.Gate):
+    """A gate that represents a Hamiltonian from a set of Boolean functions."""
+
+    def __init__(
+        self,
+        qubit_map: Sequence[Tuple[str, int]],
+        boolean_strs: Sequence[str],
+        theta: float,
+    ):
+        """Builds a BooleanHamiltonianGate.
+
+        For each element of a sequence of Boolean expressions, the code first transforms it into a
+        polynomial of Pauli Zs that represent that particular expression. Then, we sum all the
+        polynomials, thus making a function that goes from a series to Boolean inputs to an integer
+        that is the number of Boolean expressions that are true.
+
+        For example, if we were using this gate for the unweighted max-cut problem that is typically
+        used to demonstrate the QAOA algorithm, there would be one Boolean expression per edge. Each
+        Boolean expression would be true iff the vertices on that are in different cuts (i.e. it's)
+        an XOR.
+
+        Then, we compute exp(-j * theta * polynomial), which is unitary because the polynomial is
+        Hermitian.
+
+        Args:
+            boolean_strs: The list of Sympy-parsable Boolean expressions.
+            qubit_map: map of string (boolean variable name) to qubit dimension.
+            theta: The evolution time (angle) for the Hamiltonian
+        """
+        self._qubit_map: Sequence[Tuple[str, int]] = qubit_map
+        self._boolean_strs: Sequence[str] = boolean_strs
+        self._theta: float = theta
+
+    def _qid_shape_(self):
+        return tuple(dim for _, dim in self._qubit_map)
+
+    def on(self, *qubits) -> 'cirq.Operation':
+        return gate_operation.GateOperation(self, qubits)
+
+    def _value_equality_values_(self):
+        return self._qubit_map, self._boolean_strs, self._theta
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            'cirq_type': self.__class__.__name__,
+            'qubit_map': self._qubit_map,
+            'boolean_strs': self._boolean_strs,
+            'theta': self._theta,
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, qubit_map, boolean_strs, theta, **kwargs):
+        return cls(qubit_map, boolean_strs, theta)
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        qubit_map = dict(zip([name for name, _ in self._qubit_map], qubits))
+        boolean_exprs = [sympy_parser.parse_expr(boolean_str) for boolean_str in self._boolean_strs]
```

Comment:
Would it make sense to slightly refactor the code to share with the other _decompose_ function above?

Reply:
If we decide to keep both, yes. I'm hoping we can deprecate the operation though since gates are more flexible.